python/lsst/daf/butler/_butler.py

```diff
@@ -649,14 +648,21 @@ def remove(self, datasetRefOrType, dataId=None, *, delete=True, remember=True, *
         dataId = self.registry.expandDataId(dataId, graph=datasetType.dimensions, **kwds)
         ref = self.registry.find(self.collection, datasetType, dataId)
         if delete:
-            for r in itertools.chain([ref], ref.components.values()):
-                # If dataset is a composite, we don't know whether it's the
-                # parent or the components that actually need to be removed,
-                # so try them all and swallow errors.
-                try:
+            # There is a difference between a concrete composite and virtual
+            # composite. In a virtual composite the datastore is never
+            # given the top level DatasetRef. In the concrete composite
```

Comment:
Does this mean there is a difference in behavior at the butler level when we try to delete different kinds of composites?  Or is it just butler's responsibility rather than datastore's to iterate over components in the virtual composite case?

Reply:
In one case butler sends the composite ref to the datastore and datastore stores its own entries because it is writing a single file. In the virtual case (I still get these the wrong way round) the butler keeps the top level composite ref to itself and only sends the components to the datastore.  Previously we were doing a "send all refs regardless" approach and mostly things were throwing exceptions because either the ref was never known to butler, or else after the first ref was removed the component refs were already deleted. Now at least the two paths are separate just as in getDirect.